File: frontend/models/preparation-material.js
import PropTypes from 'prop-types';
import { attr, fk } from 'redux-orm';

import BaseModel from '.';

class PreparationMaterial extends BaseModel {}

PreparationMaterial.modelName = 'PreparationMaterial';

PreparationMaterial.fields = {
  ...BaseModel.fields,
  id: attr(),
  position: attr(),
  title: attr(),
  subtitle: attr(),
  description: attr(),
  mediumType: attr(),
  icon: attr(),
  link: attr(),
  topic: fk('Topic', 'preparationMaterials'),
};

PreparationMaterial.defaultProps = {
  ...BaseModel.defaultProps,
  id: null,
  position: null,
  title: '',
  subtitle: '',
  description: '',
  mediumType: null,
  icon: null,
  link: null,
};

const Shape = PropTypes.shape({
  id: PropTypes.string,
  position: PropTypes.number,
  title: PropTypes.string,
  subtitle: PropTypes.string,
  description: PropTypes.string,
  mediumType: PropTypes.string,
  icon: PropTypes.string,
  link: PropTypes.string,
});

export default PreparationMaterial;
export { Shape };

File: spec/factories/preparation_material.rb
# frozen_string_literal: true

require 'faker'

FactoryBot.define do
  factory :preparation_material do
    topic

    sequence(:position)

    title     { Faker::Lorem.words(number: 6).join ' ' }
    subtitle  { Faker::Lorem.words(number: 6).join ' ' }
    link      { Faker::Internet.url }
    published { true }

    medium_type do
      medium_types = PreparationMaterial.medium_types.keys
      medium_types[rand(medium_types.length - 1)]
    end

    icon do
      case medium_type.to_sym
      when :medium_type_video
        'play'
      when :medium_type_document
        'file-o-pdf'
      when :medium_type_link
        'link'
      else
        'question'
      end
    end

    trait :unpublished do
      published { false }
    end
  end
end
